pkg/scheduler/framework/plugins/volumerestrictions/volume_restrictions.go

```diff
@@ -241,3 +291,81 @@ func New(_ runtime.Object, handle framework.Handle, fts feature.Features) (frame
 		enableReadWriteOncePod: fts.EnableReadWriteOncePod,
 	}, nil
 }
+
+// canBePreempt only verifies the volume, not the resource/affinity..
+func canBePreempt(nodeInfo *framework.NodeInfo, key string) bool {
+	namespace, pvcName := strings.Split(key, "/")[0], strings.Split(key, "/")[1]
+	anotherNodeInfo := nodeInfo.Clone()
+	for _, podInfo := range anotherNodeInfo.Pods {
+		pod := podInfo.Pod
+		if pod.Namespace != namespace {
+			continue
+		}
+		for _, vol := range pod.Spec.Volumes {
+			if vol.PersistentVolumeClaim.ClaimName == pvcName {
+				return true
+			}
+		}
+	}
+	return false
+}
+
+// preFilterState computed at PreFilter and used at Filter.
+type preFilterState struct {
+	nodes []*v1.Node
+	// If the AccessModes of the pvc used by a pod is ReadWriteOncePod,
+	// pod's namespace and pvcName are added to pvck
+	pvck sets.String
+}
+
+func (s *preFilterState) Clone() framework.StateData {
+	return s
+}
+
+func getPreFilterState(cycleState *framework.CycleState) (*preFilterState, error) {
+	c, err := cycleState.Read(preFilterStateKey)
+	if err != nil {
+		// preFilterState doesn't exist, likely PreFilter wasn't invoked.
+		return nil, fmt.Errorf("error reading %q from cycleState: %w", Name, err)
+	}
+
+	s, ok := c.(*preFilterState)
+	if !ok {
+		return nil, fmt.Errorf("%+v  convert to volumerestrictions.state error", c)
+	}
+	return s, nil
+}
+
+func (s *preFilterState) updateWithPod(nodeInfo *framework.NodeInfo, pvcKeys sets.String, multiplier int64) {
+	if multiplier == 1 {
+		// update nodes
+		pvcKeysNumber := len(pvcKeys)
+		for key := range pvcKeys {
+			refCount := nodeInfo.PVCRefCounts[key]
+			if refCount != 0 {
+				pvcKeysNumber--
+			}
+		}
+		if pvcKeysNumber == 0 {
+			s.nodes = append(s.nodes, nodeInfo.Node())
+		}
+		s.pvck = pvcKeys.Union(s.pvck)
```

Comment:
Aren't these keys independent of the pods running in the cluster? Initially, you just obtain them from the incoming Pod and the existing RWOP PVCs.

I'm stopping my review here. Please give the `preFilterState` variable names and comments that explain what they hold and how they change.

Reply:
I tried to modify the naming of the fields in preFilterState and add notes.

By the way, if we only need to fail quickly, then the **keysWithRWOP** field is not needed. I am a bit confused about the subsequent preemption logic (such as the condition that triggers the preemption). If the final rwop handles preemption in the same way as rwo, I will delete this filed.